Initial GraphPresenter prototype for #17

michael-hawker opened the pull request:
Fixes #17 
Fixes #27 

Pushed some things needed here to the Toolkit in [https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3033](https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3033), will update this once those get merged in.

'Done':
- [ ] ~~How to deal with paging? Use IncrementalLoadingCollection?~~ Opened Issue #54 to track for future
- [x] How to deal with QueryOptions? Can we take in XAML???
- [ ] ~~Sample: What to do with empty thumbnails, should the converter just be a 'sample' one vs. part of the lib?~~ Needs https://github.com/microsoftgraph/MSGraph-SDK-Code-Generator/issues/266 to be fixed to continue OneDrive sample
- [ ] ~~XAML Islands: Test in XAML Islands, maybe make it so XAML Islands UWP just uses MainPage from UWP sample???~~ Doesn't need to be part of this work
- [x] How do we order in reverse in Graph API??? (add 'desc' with a space before to the OrderBy field) [See here](https://docs.microsoft.com/en-us/graph/query-parameters#orderby-parameter)

TODO/Questions:
- [ ] Test: Make sure getting the right number of days in CalendarView sample
- [x] Clean-up HTML output in Teams message sample
- [x] Move FileSize Converter to Toolkit? (not worry about for now, can do later)
- [x] Remove use of NotifyTaskCompletion (is that sample only use?) Done and added work-around to #55

michael-hawker commented:
Need to fix StyleCop issues:
```
         Extensions\GraphExtensions.cs(183,1): error SA1137: Elements should have the same indentation [D:\a\1\s\Microsoft.Toolkit.Graph\Microsoft.Toolkit.Graph.csproj]
         Extensions\GraphExtensions.cs(181,47): error SA1629: Documentation text should end with a period [D:\a\1\s\Microsoft.Toolkit.Graph\Microsoft.Toolkit.Graph.csproj]
         Extensions\GraphExtensions.cs(182,38): error SA1629: Documentation text should end with a period [D:\a\1\s\Microsoft.Toolkit.Graph\Microsoft.Toolkit.Graph.csproj]
         Extensions\GraphExtensions.cs(154,32): error SA1139: Use literal suffix notation instead of casting [D:\a\1\s\Microsoft.Toolkit.Graph\Microsoft.Toolkit.Graph.csproj]
         Extensions\GraphExtensions.cs(158,40): error SA1139: Use literal suffix notation instead of casting [D:\a\1\s\Microsoft.Toolkit.Graph\Microsoft.Toolkit.Graph.csproj]
         Extensions\GraphExtensions.cs(162,40): error SA1139: Use literal suffix notation instead of casting [D:\a\1\s\Microsoft.Toolkit.Graph\Microsoft.Toolkit.Graph.csproj]
         Extensions\GraphExtensions.cs(166,40): error SA1139: Use literal suffix notation instead of casting [D:\a\1\s\Microsoft.Toolkit.Graph\Microsoft.Toolkit.Graph.csproj]
         Extensions\GraphExtensions.cs(170,40): error SA1139: Use literal suffix notation instead of casting [D:\a\1\s\Microsoft.Toolkit.Graph\Microsoft.Toolkit.Graph.csproj]
         Extensions\GraphExtensions.cs(174,40): error SA1139: Use literal suffix notation instead of casting [D:\a\1\s\Microsoft.Toolkit.Graph\Microsoft.Toolkit.Graph.csproj]

```

michael-hawker commented:
```
         Converters\FileSizeConverter.cs(11,18): error CS1591: Missing XML comment for publicly visible type or member 'FileSizeConverter' [D:\a\1\s\Microsoft.Toolkit.Graph.Controls\Microsoft.Toolkit.Graph.Controls.csproj]
         Converters\FileSizeConverter.cs(13,23): error CS1591: Missing XML comment for publicly visible type or member 'FileSizeConverter.Convert(object, Type, object, string)' [D:\a\1\s\Microsoft.Toolkit.Graph.Controls\Microsoft.Toolkit.Graph.Controls.csproj]
         Converters\FileSizeConverter.cs(23,23): error CS1591: Missing XML comment for publicly visible type or member 'FileSizeConverter.ConvertBack(object, Type, object, string)' [D:\a\1\s\Microsoft.Toolkit.Graph.Controls\Microsoft.Toolkit.Graph.Controls.csproj]
         Providers\NotifyTaskCompletion.cs(141,45): error SA1502: Element should not be on a single line [D:\a\1\s\Microsoft.Toolkit.Graph.Controls\Microsoft.Toolkit.Graph.Controls.csproj]
         Providers\NotifyTaskCompletion.cs(109,36): error SA1502: Element should not be on a single line [D:\a\1\s\Microsoft.Toolkit.Graph.Controls\Microsoft.Toolkit.Graph.Controls.csproj
```

michael-hawker commented on Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs:
```diff
@@ -0,0 +1,116 @@
+
+                // TODO: Add Exception Handling
+                // Note: CalendarView not supported https://github.com/microsoftgraph/msgraph-sdk-dotnet/issues/740
+                var response = await request.SendAsync<object>(null, CancellationToken.None).ConfigureAwait(false) as JObject;
```
Thanks Sergio, I did mean to do something here. We should probably have some thoughts in general about loading/error states for the Graph controls in general. I should at least wrap and ignore this for now though. :)

michael-hawker commented on Microsoft.Toolkit.Graph/Extensions/GraphExtensions.cs:
```diff
@@ -136,5 +137,42 @@ public static T Search<T>(this T request, string query)
+        /// </summary>
+        /// <param name="size">file size in bytes.</param>
+        /// <returns>Returns file size string.</returns>
+        public static string ToFileSizeString(this long size) // TODO: Move this to Microsoft.Toolkit
```
I completely forgot about it, I should have a reference too, I'll remove this from here. Thanks!

michael-hawker commented on SampleTest/MainPage.xaml:
```diff
@@ -25,19 +28,236 @@
+                                                        <DataTemplate x:DataType="graph:Message">
+                                                            <StackPanel>
+                                                                <wgt:PersonView PersonQuery="{x:Bind Sender.EmailAddress.Address}" ShowName="True" Margin="-4"/>
+                                                                <TextBlock Text="{Binding Subject}" Style="{StaticResource BaseTextBlockStyle}" Padding="0"/>
```
Just forgot to update it from when I was playing around the first time :)

michael-hawker commented on Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs:
```diff
@@ -0,0 +1,116 @@
+                await DispatcherHelper.ExecuteOnUIThreadAsync(() =>
+                {
+                    Content = data;
+                });
```
Ordering was odd, so couldn't apply this directly, but resolved now.

azchohfi commented on Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs:
```diff
@@ -0,0 +1,113 @@
+        /// The identifier for the <see cref="RequestBuilder"/> dependency property.
+        /// </returns>
+        public static readonly DependencyProperty RequestBuilderProperty =
+            DependencyProperty.Register(nameof(RequestBuilder), typeof(IBaseRequestBuilder), typeof(GraphPresenter), new PropertyMetadata(null));
```
What if the RequestBuilder changes after the GraphPresenter is loaded? Shouldn't this be handled on the DP's changed event?

michael-hawker replied:
https://github.com/windows-toolkit/Graph-Controls/issues/56

azchohfi commented on Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs:
```diff
@@ -0,0 +1,113 @@
+        /// <summary>
+        /// Gets or sets a string to indicate a sorting order for the <see cref="RequestBuilder"/>. This is a helper to add this specific request option to the <see cref="QueryOptions"/>.
+        /// </summary>
+        public string OrderBy { get; set; }
```
Wouldn't it make sense to make all these properties as DPs, and react (do the query again with the updated parameters) when they change?

michael-hawker replied:
For this and the RequestBuilder, I didn't expect them to change much yet. I think that's a separate PR to add that functionality in?

michael-hawker replied:
Opened issue https://github.com/windows-toolkit/Graph-Controls/issues/56 to track this for later.

azchohfi commented on Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs:
```diff
@@ -0,0 +1,113 @@
+                        data = values.ToObject(ResponseType);
+                    }
+
+                    _ = DispatcherHelper.ExecuteOnUIThreadAsync(() => Content = data);
```
Use the DispatcherQueue, not the old CoreDispatcher.

michael-hawker replied:
We're not on the 7.0.0 preview package yet. I need to update later.

michael-hawker replied:
#57

azchohfi commented on SampleTest/MainPage.xaml:
```diff
@@ -2,16 +2,19 @@
     mc:Ignorable="d"
     Background="{ThemeResource ApplicationPageBackgroundThemeBrush}"
-    xmlns:controls="using:Microsoft.Toolkit.Uwp.UI.Controls"><!-- Workaround for https://github.com/microsoft/microsoft-ui-xaml/issues/1286 -->
+    xmlns:controlsKeepThis="using:Microsoft.Toolkit.Uwp.UI.Controls"><!-- Workaround for https://github.com/microsoft/microsoft-ui-xaml/issues/1286 -->
```
If we bump the target SDK, should this issue go away? If so, we should do this.

azchohfi replied:
Also, this is being added right here on the top(line 13), so I don't think you need it anymore.

michael-hawker replied:
I might bump/test that as a separate PR?

michael-hawker replied:
Added to #57

azchohfi commented on Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/QueryOption.cs:
```diff
@@ -0,0 +1,34 @@
+    /// <summary>
+    /// XAML Proxy for <see cref="Microsoft.Graph.QueryOption"/>.
+    /// </summary>
+    public sealed class QueryOption
```
Shouldn't we reuse the existing ListDictionary and keep from adding another Key<->Value mapping class?
Something like this:
```xaml
...
xmlns:specialized="using:System.Collections.Specialized"
...
<specialized:ListDictionary>
    <x:String x:Key="key1">Value1</x:String>
    <x:String x:Key="key2">Value2</x:String>
</specialized:ListDictionary>
```
One less class to maintain.

michael-hawker replied:
I didn't know this existed 🙂. I do like having close mapping to the intent though for the Graph. And we'd lose the implicit operator converter, eh? Let's chat quick in the morning.

michael-hawker replied:
I don't know how we bind to Value in these cases, so let's leave this for now and explore more later. I think there does need to be some refactoring to ensure timings and/or dynamic updates to query options...

michael-hawker commented on Microsoft.Toolkit.Graph/Providers/MockProvider.cs:
```diff
@@ -42,22 +44,15 @@ private set
+                        var requestUri = requestMessage.RequestUri.ToString();
+
+                        // Prepend Proxy Service URI to our request
+                        requestMessage.RequestUri = new Uri(GRAPH_PROXY_URL + HttpUtility.UrlEncode(requestUri));
```
```suggestion
                        requestMessage.RequestUri = new Uri(GRAPH_PROXY_URL + Uri.EscapeDataString(requestUri));
```

Looks like we can use `Uri.EscapeDataString` here instead which reduces one more reference (System.Web) and is more compatible with the Uno build.

michael-hawker commented on SampleTest/MainPage.xaml.cs:
```diff
@@ -11,9 +16,44 @@ namespace SampleTest
+
+        public static string ToLocalTime(DateTimeTimeZone value)
+        {
+            //// Workaround for https://github.com/microsoft/microsoft-ui-xaml/issues/2407
```
Made `//` instead

michael-hawker commented:
@azchohfi thanks for the review, addressed your comments as well and opened #56 for tracking. If it looks good, please merge this.

Also added #57 for updating to our new 7.0.0-preview2 package and the Target SDK bump. If you have a chance to do that update today, that'd be big help for UnoConf. Then we could update our preview package for this as well to align with `7.0.0-preview2`.

Thanks! 🦙❤

azchohfi commented:
Working on it!
Small issue!
```
         Controls\GraphPresenter\QueryOption.cs(28,57): error SA1629: Documentation text should end with a period [D:\a\1\s\Microsoft.Toolkit.Graph.Controls\Microsoft.Toolkit.Graph.Controls.csproj]
```

azchohfi commented on Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/QueryOption.cs:
```diff
@@ -0,0 +1,34 @@
+        /// <summary>
+        /// Implicit conversion for <see cref="QueryOption"/> to <see cref="Microsoft.Graph.QueryOption"/>.
+        /// </summary>
+        /// <param name="option">query option to convert</param>
```
```suggestion
        /// <param name="option">query option to convert.</param>
```
